api/src/main/java/org/apache/iceberg/expressions/Literals.java

```diff
@@ -426,23 +427,55 @@ protected Type.TypeID typeId() {
   }
 
   static class TimestampLiteral extends ComparableLiteral<Long> {
-    TimestampLiteral(Long value) {
+    private final TimestampType.Unit unit;
+
+    TimestampLiteral(TimestampType.Unit unit, Long value) {
       super(value);
+      this.unit = unit;
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public <T> Literal<T> to(Type type) {
       switch (type.typeId()) {
         case TIMESTAMP:
-          return (Literal<T>) this;
+          TimestampType.Unit toUnit = ((TimestampType) type).unit();
```

Comment:
The decision to use the same `TimestampType` class and use `unit` makes this logic more complicated than necessary.

The idea behind having a Java enum, `Type.TypeID`, was that in most cases you can use a switch statement and very simple logic within that switch. The reason why `TimestampType` was parameterized by a boolean for `adjustToUTC` is that the logic in most places is identical for `timestamp` and `timestamptz` because they have the same internal representation.

For `timestamp_ns` and `timestamptz_ns`, I don't think that the same logic applies. The internal representation is different between nanos and micros, so you end up with this complexity, where you have to add a switch on a different enum. I think it would be simpler to have a `TimestampNanosType` class instead of `Unit`.

Reply:
Done.